### PR TITLE
Crypto backend send receive approve

### PR DESCRIPTION
**Pull request by kachwalUttamsharma**

### CODE REVIEW

**Comment by kachwalUttamsharma** [comment]
Screenshot 2021-02-19 at 10 06 41 PM

**Comment by ankurnarkhede** [comment]
@kachwalUttamsharma , is there any contract for this?

cc: @swarajpure

**Comment by kachwalUttamsharma** [comment]
Yes, I believe. @Kratika0907 mite have created it.

**Comment by kachwalUttamsharma** [comment]
> @kachwalUttamsharma , is there any contract for this?
> 
> cc: @swarajpure 

Yes, I believe. @Kratika0907 mite have created it.

**Comment by ankurnarkhede** [comment]
@kachwalUttamsharma , @Kratika0907 , can you link it here, along with the issue?

**Comment by kachwalUttamsharma** [comment]
> @kachwalUttamsharma , @Kratika0907 , can you link it here, along with the issue?

https://github.com/Real-Dev-Squad/website-api-contracts/pull/10

File: controllers/cryptoController.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+// }
+const send = async (req, res) => {
+  try {
+    const request = req.body
+    const [userTo, userFrom, currency, amount] = [request.to, request.from, request.currency, request.amount]
+    const message1 = `${amount} ${currency} Coins Transfered From ${userFrom} to ${userTo}`
+    const verifyUserTO = await usersdata.fetchUser(userTo)
+    const verifyUserFrom = await usersdata.fetchUser(userFrom)
+    if (verifyUserTO && verifyUserFrom) {
+      if (currency === 'brass' && verifyUserFrom.user.coins.brass >= amount) {
```
Comment (Kratika0907) [inline controllers/cryptoController.js:24]: This logic can be abstracted into a different function.

File: controllers/cryptoController.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    const request = req.body
+    const [notification, userFrom] = [request.notification, request.userName]
+    const verifyUserFrom = await usersdata.fetchUser(userFrom)
+    const message = notification.split(' ')
+    const [amount, currency, userTO] = [message[0], message[1], message[(message.length) - 1]]
+    const verifyUserTo = await usersdata.fetchUser(userTO)
+    // const message1 = `${amount} ${currency} Coins Transfered From ${userFrom} to ${userTo}`;
+    if (verifyUserTo && verifyUserFrom) {
+      if (currency === 'brass' && verifyUserFrom.user.coins.brass >= amount) {
+        verifyUserFrom.user.coins.brass = verifyUserFrom.user.coins.brass - amount
```
Comment (Kratika0907) [inline controllers/cryptoController.js:107]: Same here. Redundant code

File: routes/index.js [hunk 3]
```diff
@@ -9,5 +9,6 @@ app.use('/tasks', require('./tasks'))
 app.use('/challenges', require('./challenges.js'))
 app.use('/pullrequests', require('./pullrequests.js'))
 app.use('/contributions', require('./contributions'))
+app.use('/cryptoTransaction', require('./crypto.js'))
```
Comment (Kratika0907) [inline routes/index.js:12]: Joi validations are missing.

File: controllers/cryptoController.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+/**
+ * Receive Money from one User to Other
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+
+// http://localhost:3000/cryptoTransaction/approvedRequest
+// content-type: appliction/json -> header
+// Json Body
```
Comment (Kratika0907) [inline controllers/cryptoController.js:91]: Keep comments minimum. Just a higher overview of what is enough.

**Review by Kratika0907 — CHANGES_REQUESTED** [review]

File: controllers/cryptoController.js (hunk 1 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:24]: ok Working on it

File: controllers/cryptoController.js (hunk 2 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:107]: will modify it

File: controllers/cryptoController.js (hunk 4 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:91]: ok

File: routes/index.js (hunk 3 above)
Comment (kachwalUttamsharma) [reply routes/index.js:12]: added joi validation.

File: controllers/cryptoController.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const send = async (req, res) => {
+  try {
+    const request = req.body
+    const [userTo, userFrom, currency, amount] = [request.to, request.from, request.currency, request.amount]
+    const message1 = `${amount} ${currency} Coins Transfered From ${userFrom} to ${userTo}`
+    const verifyUserTO = await usersdata.fetchUser(userTo)
+    const verifyUserFrom = await usersdata.fetchUser(userFrom)
```
Comment (Kratika0907) [inline controllers/cryptoController.js:14]: use Promise.all api to fetch both things and then await

File: controllers/cryptoController.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    const resultObj = calculation.cryptoCalc(verifyUserTO, verifyUserFrom, currency, amount)
+    return res.json({
+      message: message1,
```
Comment (Kratika0907) [inline controllers/cryptoController.js:17]: Proper message name like success Message or something related

File: controllers/cryptoController.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  try {
+    const request = req.body
+    const [userTo, userFrom, currency, amount] = [request.to, request.from, request.currency, request.amount]
+    const notificationMessage = `${amount} ${currency} Coins Requested By ${userFrom}`
+    const message1 = `${amount} ${currency} Coins Requested By ${userFrom} From ${userTo}`
+    const verifyUserTO = await usersdata.fetchUser(userTo)
+    const verifyUserFrom = await usersdata.fetchUser(userFrom)
+    if (verifyUserFrom && verifyUserTO) {
+      const notificationArray = verifyUserTO.user.notification
+      notificationArray.push(notificationMessage)
```
Comment (Kratika0907) [inline controllers/cryptoController.js:42]: Does this push creates a new  notification in db  also.

File: controllers/cryptoController.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    const verifyUserTo = await usersdata.fetchUser(userTO)
+    // const message1 = `${amount} ${currency} Coins Transfered From ${userFrom} to ${userTo}`;
+    const resultObj = calculation.cryptoCalc(verifyUserTo, verifyUserFrom, currency, amount)
+    return res.json({
+      message: notification,
+      to: resultObj.verifyUserTO,
+      from: resultObj.verifyUserFrom
+    })
+  } catch (error) {
+    logger.error(`Error while processing pull requests: ${error}`)
```
Comment (Kratika0907) [inline controllers/cryptoController.js:81]: Why this error

**Review by Kratika0907 — CHANGES_REQUESTED** [review]

File: controllers/cryptoController.js (hunk 7 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:42]: Yes I have attached screenshot of collection. we have notification array.

File: controllers/cryptoController.js (hunk 6 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:17]: ok
will implement that.

File: controllers/cryptoController.js (hunk 5 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:14]: ok working on it.

File: controllers/cryptoController.js (hunk 8 above)
Comment (kachwalUttamsharma) [reply controllers/cryptoController.js:81]: for logging purpose in console for debugging purppose.

**Pull request closed without merging**